### PR TITLE
Added support for custom urls, an explicit work dir, and better logs

### PR DESCRIPTION
**Pull request by UebelAndre**
Hey, @ehuss I'm wondering if you'd be willing to accept these changes into this tool. I found the crate to be incredibly useful but it lacked the control (as a library) that I was hoping to get. This PR wraps the existing logic for cloning into a struct called`Cloner`. The API is the same as before but now this object can be broken out and used in cases where hosts are behind some firewall and need to use proxied addresses. These changes unblock the work I'm trying to do for 

Here you can see the binary built in this project script continues to work
```output
cargo run --bin cargo-clone clone --help
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
     Running `target/debug/cargo-clone clone --help`
```
```output
cargo-clone-clone 0.1.2
Clone a package from crates.io.

USAGE:
    cargo-clone clone [OPTIONS] <name> [extra]...

FLAGS:
    -h, --help    Prints help information

OPTIONS:
        --method <method>      Method to fetch package. [default: auto]  [possible values: crate,
                               git, hg, pijul, fossil, auto]
        --version <version>    Version to download.

ARGS:
    <name>        Package name to clone.
    <extra>...    Additional arguments passed to clone command.
```

Here is the binary actually pulling data.
```output
cargo run --bin cargo-clone clone cargo-raze --version 0.5.0
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
     Running `target/debug/cargo-clone clone cargo-raze --version 0.5.0`
```
```output
Downloading `https://crates.io/api/v1/crates/cargo-raze/0.5.0/download`
cargo-raze-0.5.0/.cargo_vcs_info.json
cargo-raze-0.5.0/.rustfmt.toml
cargo-raze-0.5.0/Cargo.lock
cargo-raze-0.5.0/Cargo.toml
cargo-raze-0.5.0/Cargo.toml.orig
cargo-raze-0.5.0/README.md
cargo-raze-0.5.0/src/bazel.rs
cargo-raze-0.5.0/src/bin/cargo-raze.rs
cargo-raze-0.5.0/src/context.rs
cargo-raze-0.5.0/src/lib.rs
cargo-raze-0.5.0/src/license.rs
cargo-raze-0.5.0/src/metadata.rs
cargo-raze-0.5.0/src/planning.rs
cargo-raze-0.5.0/src/rendering.rs
cargo-raze-0.5.0/src/settings.rs
cargo-raze-0.5.0/src/templates/crate.BUILD.template
cargo-raze-0.5.0/src/templates/partials/build_script.template
cargo-raze-0.5.0/src/templates/partials/common_attrs.template
cargo-raze-0.5.0/src/templates/partials/remote_crates_patch.template
cargo-raze-0.5.0/src/templates/partials/rust_binary.template
cargo-raze-0.5.0/src/templates/partials/rust_library.template
cargo-raze-0.5.0/src/templates/remote_crates.bzl.template
cargo-raze-0.5.0/src/templates/workspace.BUILD.template
cargo-raze-0.5.0/src/util.rs
cargo-raze-0.5.0/test_fixtures/metadata.txt
```

The passing tests can be seen here:
```output
    Finished test [unoptimized + debuginfo] target(s) in 0.09s
     Running target/debug/deps/test_clone-e84dcfa18b39d018

running 9 tests
test err_both_new_old_style_version ... ok
test parse_version_err ... ok
test parse_version_empty ... ok
test unknown_crate ... ok
test extra_args_crate ... ok
test version_with_method ... ok
Cloning into 'bf'...
remote: Enumerating objects: 27, done.
remote: Counting objects: 100% (27/27), done.
remote: Compressing objects: 100% (26/26), done.
test parse_version_req_ok ... ok
remote: Total 27 (delta 1), reused 13 (delta 0), pack-reused 0
Unpacking objects: 100% (27/27), done.
test clone_git_args ... ok
test clone_fossil ... test clone_fossil has been running for over 60 seconds
Round-trips: 2   Artifacts sent: 0  received: 2983
Clone done, sent: 581  received: 1874194  ip: 74.208.146.128
Rebuilding repository meta-data...
  100.0% complete...
Extra delta compression...
Vacuuming the database...
project-id: 077a78b0b7daff375cdcfb1dccaa8343603c92ea
server-id:  2d9808a01a63081e3029b5d7af1347df43c096ac
admin-user: user (password is "**********")
test clone_fossil ... ok

test result: ok. 9 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests cargo_clone

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

### CODE REVIEW

**Comment by UebelAndre** [comment]
Hey @ehuss, friendly ping on this PR. Wondering if you're interested in accepting PRs and/or if you have the time 😅

**Comment by UebelAndre** [comment]
Hey, added a bit more here and improved the docs, just FYI 😄 

Tests are all still passing.

File: tests/test_clone.rs [hunk 1]
```diff
@@ -70,8 +70,16 @@ fn unknown_crate() {
...
 #[test]
 fn clone_fossil() {
+    // Ensure `fossil` is a spawnable process on the machine
+    assert!(std::process::Command::new("fossil").output().is_ok());
+
     clone("fossil", "rs-graph", None, &["graph.fossil"]).unwrap();
     assert_downloaded("graph.fossil");
+
+    // Also delete the *-journal file
+    if Path::new("./graph.fossil-journal").exists() {
```
Comment (ehuss) [inline tests/test_clone.rs:80]: Can you say why this is needed? I don't see it being created on my system.

File: src/main.rs [hunk 2]
```diff
@@ -52,11 +78,17 @@ fn main() {
         .values_of("extra")
         .map_or_else(Vec::new, |e| e.collect());
 
-    let result = cargo_clone::clone(method, name, version, &extra);
+    let cloner = cargo_clone::Cloner::default().unwrap();
+    let result = cloner.clone(
+        cargo_clone::CloneMethodKind::from(method).unwrap(),
```
Comment (ehuss) [inline src/main.rs]: Can this include a comment why the `unwrap` is ok?

File: src/lib.rs [hunk 3]
```diff
@@ -1,70 +1,85 @@
...
 
-    Ok(())
+    /// Defaults to https://gitlab.com
+    gitlab_url: String,
+
+    /// Defaults to https://api.bitbucket.org/2.0/repositories
+    bitbutcket_url: String,
+
+    /// Defaults to `std::env::current_dir()`
+    work_dir: PathBuf,
```
Comment (ehuss) [inline src/lib.rs]: Can this comment explain what the value is for?  Any maybe a name like `out_dir` would be more descriptive?

File: src/main.rs [hunk 4]
```diff
@@ -52,11 +78,17 @@ fn main() {
         .values_of("extra")
         .map_or_else(Vec::new, |e| e.collect());
 
-    let result = cargo_clone::clone(method, name, version, &extra);
+    let cloner = cargo_clone::Cloner::default().unwrap();
```
Comment (ehuss) [inline src/main.rs]: It seems like this will panic if it can't determine the current dir.

File: src/lib.rs [hunk 5]
```diff
@@ -210,88 +119,324 @@ fn get_repo(pkg_info: &Value) -> Result<Option<String>, Error> {
...
+        if spec_version_req.is_some() && version.is_some() {
+            bail!("Cannot specify both a :version and --version.");
+        }
+        let version_req = version
+            .or(spec_version_req)
+            .map(check_semver_req)
+            .transpose()?;
+        let pkg_info = self.get_pkg_info(name)?;
+        let repo = get_repo(&pkg_info)?;
+        let (method, repo) = match method_kind.command() {
```
Comment (ehuss) [inline src/lib.rs]: Perhaps this should just match on the enum?

**Review by ehuss — COMMENTED** [review]
Can you say more about why you need the ability to customize each URL?  You mentioned proxies, but usually proxy configuration is more of a global thing.

Can you say more about why the change to use `env_logger` for the output?

File: src/main.rs (hunk 4 above)
Comment (UebelAndre) [reply src/main.rs]: That's true but is this different behavior from what [Command](https://doc.rust-lang.org/std/process/struct.Command.html) will do? I can make `out_dir` (previously `work_dir` `Option<PathBuf>` and if it's not specified it's not set. Is that an improvement?

File: tests/test_clone.rs (hunk 1 above)
Comment (UebelAndre) [reply tests/test_clone.rs:80]: Odd, I see it on mine. I'm running on MacOS 10.15.6 and I installed Fossil via Brew. I know very little about Fossil but this file gets generated every run for me.

File: src/lib.rs (hunk 5 above)
Comment (UebelAndre) [reply src/lib.rs]: Done

File: src/lib.rs (hunk 3 above)
Comment (UebelAndre) [reply src/lib.rs]: Added a small comment and changed the name.

File: src/main.rs (hunk 2 above)
Comment (UebelAndre) [reply src/main.rs]: Added a comment about line 83

**Comment by UebelAndre** [comment]
> Can you say more about why the change to use `env_logger` for the output?

My primary use of this Crate is as a library within another CLI tool. I'd like to be able to control what gets written and what doesn't.  Because of this I switched the `println!` statements to use `log` and in order to maintain the previous behavior I used `env_logger` to print things the way they used to,

**Comment by UebelAndre** [comment]
> Can you say more about why you need the ability to customize each URL? You mentioned proxies, but usually proxy configuration is more of a global thing.

~~I don't think I understand what you mean by proxy configurations being a global thing~~. In my experience, if you're behind some corporate firewall, a unique address is assigned to public sites. These locations can often double as a "pull-through" or "caching" server.

Edit: Sorry, needed my morning coffee to think. That makes sense.

**Comment by UebelAndre** [comment]
@ehuss Thanks so much for the review! I have some questions, answered some questions, and addressed some feedback. Ready for you to take another look whenever you get a chance 😄

File: src/lib.rs [hunk 6]
```diff
@@ -1,70 +1,98 @@
...
-                bail!(
-                    "Specifying a version `{}` only works with the `crate` method.",
-                    version_req
-                );
-            }
-            run_clone(method, &repo, extra)?;
+    }
+}
+
+impl std::fmt::Display for CloneMethodKind {
```
Comment (ehuss) [inline src/lib.rs]: This display impl doesn't seem to be really necessary.

File: src/lib.rs [hunk 7]
```diff
@@ -210,88 +132,327 @@ fn get_repo(pkg_info: &Value) -> Result<Option<String>, Error> {
...
+            | CloneMethodKind::Fossil => {
+                if let Some(version_req) = version_req {
+                    bail!(
+                        "Specifying a version `{}` only works with the `crate` method.",
+                        version_req
+                    );
+                }
+                self.run_clone(method.command(), &repo, extra)?;
+            }
+            _ => bail!("Unsupported method `{}`", method),
```
Comment (ehuss) [inline src/lib.rs]:
```suggestion
            CloneMethodKind::Auto => unreachable!(),
```

File: src/main.rs (hunk 4 above)
Comment (ehuss) [reply src/main.rs]: I was just thinking it could print a more friendly error instead of panicking (like the code previously did).

**Comment by ehuss** [comment]
Regarding the proxy, I guess my experience is that there is usually a single proxy address, and then all HTTP traffic goes through that (or, it is a passive MITM proxy).  I've never seen one that has a different address for each site, though I guess that doesn't mean they don't exist.

File: src/main.rs (hunk 4 above)
Comment (UebelAndre) [reply src/main.rs]: Updated

File: src/lib.rs (hunk 6 above)
Comment (UebelAndre) [reply src/lib.rs]: Updated

File: src/lib.rs (hunk 7 above)
Comment (UebelAndre) [reply src/lib.rs]: Applied this change manually, I didn't know about `unreachable`, pretty cool! 😄

File: tests/test_clone.rs (hunk 1 above)
Comment (UebelAndre) [reply tests/test_clone.rs:80]: I see references to the `-journal` files in some posts but am having difficulty finding documentation on it.

https://fossil-scm.org/home/doc/trunk/www/server/any/cgi.md
https://fossil-scm.org/forum/forumpost/2c7ccb6b04

It seems to be a thing but again, I don't have any familiarity with that scm tool so I don't know why I'd have a `*-journal` file and you wouldn't

**Comment by UebelAndre** [comment]
@ehuss I'm committing new commits but can squash them down if necessary. Thought things would be easier to review this way.

Regarding the proxy, I feel having the ability to specify the url's would be extremely beneficial. I can't speak to that being good or common practice or anything but I've come across situations where embedded URLs became problematic. 

I've addressed your feedback and I think this is ready for another review 😄

File: tests/test_clone.rs (hunk 1 above)
Comment (UebelAndre) [reply tests/test_clone.rs:80]: I can remove this if you prefer that. Just let me know.

Comment (ehuss) [reply tests/test_clone.rs:80]: Nah. I have no idea why it is showing up for you, but it doesn't harm anything.

**Review by ehuss — APPROVED** [review]
Thanks!

**Comment by UebelAndre** [comment]
@ehuss YAAAYYY!! Thank you so much!! Could I also get you to do a new release to crates.io? 😀

**Comment by ehuss** [comment]
@UebelAndre Hm, after looking at the API a little more closely, I decided to change it slightly.  The `new` method wasn't extensible, and was a bit awkward to use. I changed it so that each value is set separately. It's a little more verbose to set everything, but should make it easier to change things in a compatible way in the future.  Any objections?

I also switched to using tempdir for the tests to make cleanup a little easier.

**Comment by UebelAndre** [comment]
@ehuss Hey! yeah I agree with you and think the changes you made are great. No objections from me 😄 

I like the `tempfile` use in tests as well and I especially like the fact that you're setting up Github actions 👍

**Comment by ehuss** [comment]
@UebelAndre OK, I pushed up the new version. 🎉

**Comment by UebelAndre** [comment]
@ehuss You're my hero!!!!! Thank you so much!!! 😄 🥳 🎉

**Pull request merged**